src/io/http/src/main/scala/Clients.scala

```diff
@@ -0,0 +1,346 @@
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+class BatchIterator[T](val it: Iterator[T],
+                       maxNum: Int = Integer.MAX_VALUE)
+  extends Iterator[List[T]] {
+
+  val queue: BlockingQueue[T] = new LinkedBlockingQueue[T](maxNum)
+  var hasStarted = false
+  val finishedLatch = new CountDownLatch(1)
+
+  private val thread: Thread = new Thread {
+    override def run(): Unit = {
+      while(it.synchronized(it.hasNext)){
+        val datum = it.synchronized(it.next())
+        queue.put(datum)
+      }
+      finishedLatch.countDown()
+    }
+  }
+
+  override def hasNext: Boolean = {
+    if (!hasStarted) {
+      it.hasNext
+    } else {
+      it.synchronized(it.hasNext) ||
+      !queue.isEmpty || {
+        finishedLatch.await()
+        !queue.isEmpty
+      }
+      // Final clause needed to ensure the fetching thread
+      // can finish before the iterator is exhausted.
+      // This blocking should be kept in the final clause
+      // To optimize performance
+    }
+  }
+
+  def start(): Unit = {
+    hasStarted = true
+    thread.start()
+  }
+
+  def close(): Unit = {
+    thread.interrupt()
+  }
+
+  override def next(): List[T] = {
+    if (!hasStarted) {
+      start()
+    }
+    assert(hasNext)
+
+    val results = new java.util.ArrayList[T]()
+    queue.drainTo(results)
+    if (results.isEmpty) {
+      List(queue.take())
+    } else {
+      results.toList
+    }
+  }
+}
+
+private[ml] trait ClientTypeAliases {
+  protected type Context = mutable.Map[String, Any]
+
+  case class Response(response: CloseableHttpResponse, context: Context) {
+    def this(response: CloseableHttpResponse) = this(response, mutable.Map())
+  }
+
+  case class Request(request: HttpUriRequest, context: Context) {
+    def this(request: HttpUriRequest) = this(request, mutable.Map())
+  }
+
+  protected type Client = CloseableHttpClient
+}
+
+private[ml] trait BaseClient[In, Out]
+  extends AutoCloseable with ClientTypeAliases {
+
+  protected def sendWithRetries(request: Request,
+                                retriesLeft: Seq[Long]
+                               ): Response = {
+    val originalResponse = Response(client.execute(request.request), request.context)
+    val response = handle(originalResponse)
+    response.getOrElse({
+      if (retriesLeft.isEmpty) {
+        throw new IllegalArgumentException(
+          s"No longer retrying. " +
+            s"Code: ${originalResponse.response.getStatusLine.getStatusCode}, " +
+            s"Message: $originalResponse")
+      }
+      Thread.sleep(retriesLeft.head)
+      sendWithRetries(request, retriesLeft.tail)
+    })
+  }
+
+  protected def handle(responseWithContext: Response): Option[Response] = {
+    val response = responseWithContext.response
+    val code = response.getStatusLine.getStatusCode
+    if (code == 429) {
```

Comment:
Need a way to customize error handling. Hard-coding this logic seems a bit inflexible but I don't have better suggestions unfortunately :-(